From 95738afdd25cc70f5dbf13df2877cbf18ce97e38 Mon Sep 17 00:00:00 2001
From: Morten Rasmussen <morten.rasmussen@arm.com>
Date: Wed, 30 Mar 2016 14:20:12 +0100
Subject: [PATCH 58/78] sched/fair: Add energy_diff dead-zone margin

It is not worth the overhead to migrate tasks for tiny insignificant
energy savings. To prevent this, an energy margin is introduced in
energy_diff() which effectively adds a dead-zone that rounds tiny energy
differences to zero. Since no scale is enforced for energy model data
the margin can't be absolute. Instead it is defined as +/-1.56% energy
saving compared to the current total estimated energy consumption.
---
 kernel/sched/fair.c | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index 50dbbbd..6d24973 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -5358,6 +5358,7 @@ static int energy_diff(struct energy_env *eenv)
 	struct sched_domain *sd;
 	struct sched_group *sg;
 	int sd_cpu = -1, energy_before = 0, energy_after = 0;
+	int diff, margin;
 
 	struct energy_env eenv_before = {
 		.util_delta	= 0,
@@ -5390,7 +5391,18 @@ static int energy_diff(struct energy_env *eenv)
 		}
 	} while (sg = sg->next, sg != sd->groups);
 
-	return energy_after-energy_before;
+	/*
+	 * Dead-zone margin preventing too many migrations.
+	 */
+
+	margin = energy_before >> 6; /* ~1.56% */
+
+	diff = energy_after-energy_before;
+
+	if (abs(diff) < margin)
+		return 0;
+
+	return diff;
 }
 
 /*
-- 
2.7.4

